apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts

```
@@ -0,0 +1,90 @@
+import { GetPreferences, GetPreferencesCommand, GetPreferencesResponseDto } from '@novu/application-generic';
+
+import { GetWorkflowCommand } from './get-workflow.command';
+import { NotificationTemplateMapper } from '../../mappers/notification-template-mapper';
+import { WorkflowNotFoundException } from '../../exceptions/workflow-not-found-exception';
+import { WorkflowResponseDto } from '../../dto/workflow-response-dto';
+
+@Injectable()
+export class GetWorkflowUseCase {
+  constructor(
+    private notificationTemplateRepository: NotificationTemplateRepository,
+    private controlValuesRepository: ControlValuesRepository,
+    private getPreferencesUseCase: GetPreferences
+  ) {}
+  async execute(command: GetWorkflowCommand): Promise<WorkflowResponseDto> {
+    const notificationTemplateEntity = await this.notificationTemplateRepository.findByIdQuery({
+      id: command._workflowId,
+      environmentId: command.user.environmentId,
+    });
+
+    if (!notificationTemplateEntity) {
+      throw new WorkflowNotFoundException(command._workflowId);
+    }
+    const stepIdToControlValuesMap = await this.getControlsValuesMap(notificationTemplateEntity.steps, command);
+    const preferences = await this.getPreferencesForWorkflow(command);
+
+    return NotificationTemplateMapper.toResponseWorkflowDto(
+      notificationTemplateEntity,
+      preferences,
+      stepIdToControlValuesMap
+    );
+  }
+
+  private async getPreferencesForWorkflow(
+    getWorkflowCommand: GetWorkflowCommand
+  ): Promise<GetPreferencesResponseDto | undefined> {
+    const command = {
+      environmentId: getWorkflowCommand.user.environmentId,
+      organizationId: getWorkflowCommand.user.organizationId,
+      templateId: getWorkflowCommand._workflowId,
+    } as GetPreferencesCommand;
+
+    try {
+      const workflowPreferences = await this.getPreferencesUseCase.execute(GetPreferencesCommand.create(command));
+
+      return workflowPreferences;
+    } catch (e) {
```

Is this try catch required? If there is an error it must be a modelling/data issue at this stage. I'd suggest we let it bubble up.

---

it's not an issue, it's an issue with the actual use case, if you do a get before upsert it throws an exception but it's a viable use case

---

We could refactor the use case and throw expectations on the controller or middleware use case.